File: backend/authentication/src/main/java/com/hatirlabeni/authentication/service/interfaces/AuthService.java
package com.hatirlabeni.authentication.service.interfaces;

import com.hatirlabeni.authentication.dtos.mail.VerifyChangeEmailRequest;
import com.hatirlabeni.authentication.dtos.user.AuthUserResponse;
import com.hatirlabeni.authentication.dtos.google.GoogleRegisterRequest;
import com.hatirlabeni.authentication.dtos.login.LoginRequest;
import com.hatirlabeni.authentication.dtos.login.LoginResponse;
import com.hatirlabeni.authentication.dtos.login.LoginResult;
import com.hatirlabeni.authentication.dtos.mail.ChangeEmailRequest;
import com.hatirlabeni.authentication.dtos.mail.VerifyMailRequest;
import com.hatirlabeni.authentication.dtos.password.ChangePasswordRequest;
import com.hatirlabeni.authentication.dtos.password.ForgotPasswordRequest;
import com.hatirlabeni.authentication.dtos.register.RegisterRequest;
import com.hatirlabeni.authentication.dtos.register.RegisterResponse;
import com.hatirlabeni.authentication.dtos.register.ResetPasswordRequest;

import java.util.UUID;

public interface AuthService {

    // ==================================================
    // Registration & Authentication
    // ==================================================

    RegisterResponse register(RegisterRequest registerRequest);

    LoginResponse registerWithGoogle(
            GoogleRegisterRequest request,
            String email
    );

    LoginResult login(LoginRequest loginRequest);


    // ==================================================
    // Session Management
    // ==================================================

    LoginResult refresh(String token);


    // ==================================================
    // Password Management
    // ==================================================

    void forgotPassword(ForgotPasswordRequest forgotPasswordRequest);

    void resetPassword(ResetPasswordRequest resetPasswordRequest);

    void changePassword(
            String token,
            ChangePasswordRequest changePasswordRequest
    );


    // ==================================================
    // Email Management
    // ==================================================

    void createAndSendActivationToken();

    void verifyAndConsumeActivationToken(
            VerifyMailRequest verifyMailRequest
    );

    public void createAndSendChangeEmailToken(ChangeEmailRequest request);


    // ==================================================
    // User Information
    // ==================================================

    AuthUserResponse getByUuid(UUID uuid);

    AuthUserResponse getByUsername(String username);

    AuthUserResponse getUuidByEmail(String email);

    void verifyAndConsumeChangeEmailToken(VerifyChangeEmailRequest verifyChangeEmailRequest);

    // ==================================================
    // User Role Management

    // ==================================================

    void makeAdmin(UUID uuid);

    void makeUser(UUID uuid);

    // ==================================================
    // User Management

    // ==================================================

    void deleteAuthUser(UUID uuid);
}

File: frontend/src/api/AuthService.ts
import type {
  LoginRequest,
  LoginResponse,
  RegisterForm,
  RegisterResponse,
  ForgotPasswordRequest,
  ResetPasswordRequest,
  VerifyMailRequest,
  ChangePasswordRequest,
  ChangeEmailRequest,
  VerifyChangeEmailRequest,
  RefreshResponse,
  MeResponse,
} from "../types/auth";

import api from "./ApiClient";

// ─── Authentication ──────────────────────────────────────

export const register = async (
  formData: RegisterForm,
): Promise<RegisterResponse> => {
  const response = await api.post<RegisterResponse>(
    "/auth/register",
    formData,
  );

  return response.data;
};

export const login = async (
  formData: LoginRequest,
): Promise<LoginResponse> => {
  const response = await api.post<LoginResponse>(
    "/auth/login",
    formData,
  );

  return response.data;
};

export const refresh = async (): Promise<RefreshResponse> => {
  const response = await api.post<RefreshResponse>(
    "/auth/refresh",
  );

  return response.data;
};

export const logout = async (): Promise<void> => {
  await api.post("/auth/logout");
};

// ─── Current User ────────────────────────────────────────

export const me = async (): Promise<MeResponse> => {
  const response = await api.get<MeResponse>("/users/me");

  return response.data;
};

// ─── Password Management ──────────────────────────────────

export const forgotPassword = async (
  forgotPasswordRequest: ForgotPasswordRequest,
): Promise<void> => {
  await api.post(
    "/auth/forgot-password",
    forgotPasswordRequest,
  );
};

export const resetPassword = async (
  resetPasswordRequest: ResetPasswordRequest,
): Promise<void> => {
  await api.post(
    "/auth/reset-password",
    resetPasswordRequest,
  );
};

export const changePassword = async (
  changePasswordRequest: ChangePasswordRequest,
): Promise<void> => {
  await api.patch(
    "/auth/change-password",
    changePasswordRequest,
  );
};

// ─── Mail Activation ─────────────────────────────────────

export const sendMailActivationLink = async (): Promise<void> => {
  await api.post("/auth/mail-activation");
};

export const verifyMail = async (
  token: VerifyMailRequest,
): Promise<void> => {
  await api.post("/auth/mail-verify", token);
};

// ─── Email Change ────────────────────────────────────────

export const sendChangeEmailLink = async (
  changeEmailRequest: ChangeEmailRequest,
): Promise<void> => {
  await api.post(
    "/auth/change-email",
    changeEmailRequest,
  );
};

export const verifyChangeMail = async (
  verifyChangeEmailRequest: VerifyChangeEmailRequest,
): Promise<void> => {
  await api.post(
    "/auth/verify-change-email",
    verifyChangeEmailRequest,
  );
};